### DISCUSSION TITLE
C++ Programmers Tutorial/Documentation Needed

### DISCUSSION THREAD

**Discussion by ghost** [o3de/o3de#5307] [Help Wanted] [answered]
**Is your feature request related to a problem? Please describe.**

Firstly, thank you very much for all your hard work. It's a very challenging but at the same time it's enormously helpful and amazing project you've got going on. Thanks for all the effort.

After reading most of the parts under https://o3de.org/docs/api/ and https://o3de.org/docs/tools-ui/ there are some big gaps in lacking documentation that need some urgent care :) 

If we were trying to strive for a certain level on documentation, I believe Unity has some good structure and great examples on how to reach out to different levels of information based on your day to day needs in coding,

https://docs.unity3d.com/ScriptReference/index.html

The great functionality in Unity's reference guide is their search function. For example when you search for the word "Camera" all related scripting/coding documentation for that entity comes up. Additionally there's a "Switch to Manual" button present which takes you to equivalent articles from a game-engine editor perspective (not coding references).

Sessions that should be highlighted:

1. Where to expand your main game project code by creating new classes and writing code (not gems)?

2. The Doxygen documentation is very limited. Could we get a revision sometime soon to know what all public functions do, besides their names?

3. What IDE's are meant to be used (if that really matters), Visual Studio, CLion, other? The **CMakeLists.txt** files are having commands that do not contain only CMake standard commands? We are getting errors for example when I reload a CMake command and it says, **ly_get_list_relative_pal_filename** is an unknown command. Please advice in these cases how a developer can proceed with gem development outside the scope of the engine.

4. Which parts of the code should a game creator retain in a repository (whether its github, bitbucket, azure devops, etc.)? Do we really need to retain **all** these Gem cpp/cxx/hxx/h classes around?

5. When we compile, do we really need to compile the entire solution (it is a very time sinking process especially if you're making minor changes you need to quickly to test out something), or would certain parts of the solution suffice? **We would need a C++ tutorial to explain how to save time by just choosing out which stuff to be compiled without the need to recompile the entire solution to test/debug something more effectively...**

6. Can I develop Gems under my own project folder or do we have to register and code version them in the same location as other external public Gems? The project is currently spread out all over the place. For example ProjectName.GameLauncher.vcxproj file is retained in a project-centric structure under _C:\ProjectFolder\build\windows_vs2019\o3de\Code\LauncherUnified_ while the LauncherProject.cpp class within the same project is existing in an engine-centric location _C:\o3de\Code\LauncherUnified\LauncherProject.cpp_ could we please get all involved project files in one and the same location of our choice?

7. We need sessions to call out for deprecated functions/libraries and potential come-soon features/libraries coming out to replace them. There's a way for functions to be marked as deprecated (soon-to-be-replaced-with...) that could be part of Doxygen documentation (read more, https://docs.microsoft.com/en-us/cpp/cpp/deprecated-cpp?view=msvc-160) That would be enormously helpful for developers to capture deprecated areas in the code and/or mark them with //TODO: comments.

The project structure in Visual Studio 2019,
<img src="https://navaria-online.com/assets/images/game-solution.png">

Thank you in advance,

Sincerely,
Niklas Henricson

**Comment by sptramer** [comment, marked as answer]
I would recommend making this discussion part of the [Programmer guide restructuring](https://github.com/o3de/sig-docs-community/discussions/16) although it's more related to content needs. There's probably some overlap there.
